universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java

```diff
@@ -33,27 +30,28 @@ public ProgramIndexView(AdminLayout layout) {
     this.layout = layout;
   }
 
-  public Content render(ImmutableList<ProgramDefinition> programs, Http.Request request) {
+  public Content render(ActiveAndDraftPrograms programs, Http.Request request) {
     Tag contentDiv =
         div()
             .withClasses(Styles.PX_20)
             .with(
                 h1("All Programs").withClasses(Styles.MY_4),
                 each(
-                    programs.stream()
-                        .collect(new GroupByKeyCollector<>(ProgramDefinition::adminName)),
-                    e -> this.renderProgramListItem(e.getKey(), e.getValue(), request)),
+                    programs.getProgramNames(),
+                    name ->
+                        this.renderProgramListItem(
+                            programs.getActiveProgramDefinition(name),
+                            programs.getDraftProgramDefinition(name),
+                            request)),
                 renderNewProgramButton(),
                 maybeRenderPublishButton(programs, request));
 
     return layout.render(head(layout.tailwindStyles()), body(contentDiv));
   }
 
-  private Tag maybeRenderPublishButton(
-      ImmutableList<ProgramDefinition> programs, Http.Request request) {
+  private Tag maybeRenderPublishButton(ActiveAndDraftPrograms programs, Http.Request request) {
     // We should only render the publish button if there is at least one draft.
-    if (programs.stream()
-        .anyMatch(program -> program.lifecycleStage().equals(LifecycleStage.DRAFT))) {
+    if (programs.anyDraft()) {
```

Comment:
Question (blocking): if an admin creates a new draft and updates a single question, they should be able to publish that change. Will the programs that depend on that question automatically have drafts created?

Reply:
Yes - and we test that in the program lifecycle test.